keps/sig-instrumentation/2831-kubelet-tracing/README.md

```diff
@@ -0,0 +1,352 @@
+An [OTLP exporter](https://github.com/open-telemetry/opentelemetry-go/blob/main/exporters/otlp/otlptrace/otlptracegrpc/exporter.go) and
+an [OTLP trace provider](https://github.com/open-telemetry/opentelemetry-go/blob/main/sdk/trace/provider.go) along with a [context propagator](https://opentelemetry.io/docs/go/instrumentation/#propagators-and-context) will be configured.
+
+Also, designated http servers and clients will be wrapped with [otelhttp](https://github.com/open-telemetry/opentelemetry-go-contrib/tree/v0.21.0/instrumentation/net/http)
+to generate spans for sampled incoming requests and propagate context with client requests. **TODO** add specific use cases and more detail.
+
+OpenTelemetry-Go provides the [propagation package](https://github.com/open-telemetry/opentelemetry-go/blob/main/propagation/propagation.go) with which you can add custom key-value pairs known as [baggage](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/baggage/api.md). Baggage data will be propagated across services within contexts.
+
+### Running the OpenTelemetry Collector
+
+The [OpenTelemetry Collector](https://github.com/open-telemetry/opentelemetry-collector) can be run as a sidecar, a daemonset, a deployment , or a combination in which the daemonset buffers telemetry and forwards to the deployment for aggregation (e.g. tail-base sampling) and routing to a telemetry backend.  To support these various setups, the kubelet should be able to send traffic either to a local (on the control plane network) collector, or to a cluster service (on the cluster network).
+
+### Kubelet Configuration
+
+Refer to [OpenTelemetry Specification](https://github.com/open-telemetry/opentelemetry-specification/blob/ccbfdd9020ba0d581aa5d880235302f8b0eb8669/specification/resource/semantic_conventions/README.md#service)
+
+```golang
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// KubeletConfiguration contains the configuration for the Kubelet
+type KubeletConfiguration struct {
+    metav1.TypeMeta
+    ------------
+    // EnableOtelTracing enables export of opentelemetry traces
+    // +optional
+    EnableOtelTracing bool
+    // OpenTelemetryConfig
+    // +optional
+    OpenTelemetryTracingConfiguration OpenTelemetryTracingConfiguration
+}
+
+// OpenTelemetryTracingConfiguration specifies configuration for opentelemetry tracing
+type OpenTelemetryTracingConfiguration struct {
+    // +optional
+    // The ID of the service instance
+    // Defaults to node hostname. This results in distinct traces for each node.
+    TracingServiceInstanceID string
+
+    // +optional
+    // The port of the collector agent running on the node
+    // Defaults to 0.0.0.0:4317
+    CollectorEndpoint  string
+
+    // +optional
+    // SamplingRatePerMillion is the number of samples to collect per million spans.
+    // Defaults to 0.
+    SamplingRatePerMillion *int32
```

Comment:
Should this be an int or an int pointer? If it defaults to 0, is there any reason to differentiate between set and unset?

Reply:
I set to pointer to match the APIServer Configuration.